File: BioInfoPython/task-12/Motif Finding.py
# -*- coding: utf-8 -*-
"""
Created on 02-Aug-17 at 10:35 PM

@author: Pranta Sarker

Task: Motif Finding

python-3.6.2

"""         

from __future__ import print_function;
from random import *;

DNA = {
    1: 'A',
2: 'T',
3: 'C',
4: 'G'
}

motif = {};

def takeDNASequence(row, col):
    global motif;
    for i in range(row):
        for j in range(col):
            indx = randint(1 , 4);
            motif.setdefault(i, []).append(DNA[indx]);
    print("\nDNA Sequence:");
    for i in range(row):
        for j in range(len(motif[i])):
            print(motif[i][j] , end=' ');
        print("");


def Motif_Finding(lmar , row):
    lst = [];
    for i in range(row+5):
        lst.append(0);
    for i in range(len(motif[0]) - lmar + 1):
        string = "";
        sol = "";
        for j in range(i , lmar+i):
            string = string + str(motif[0][j]);
        print(string);
        instring = "";
        for k in range(1 , row):
            for l in range(0, len(motif[k]) - lmar + 1):
                instring = "";
                for p in range(l , lmar + l):
                    instring = instring + str(motif[k][p]);
                print('string = ' , string , ' ' , 'instring = ' , instring);
                if(string == instring):
                    sol = string;
                    lst[k] = 1;
        fl = True;
        for x in range(1 , row):
            if(lst[x] == 0):
                fl=False;
                break;
        if(fl == True):
            break;
        else:
            for a in range(row+5):
                lst[a] = 0;
            sol = "nai";
    #print(sol);
    if(sol == 'nai'):
        print("\nMotif did not find for this Sequences\n");
    else:
        print("\nMotif: " , sol);

print('<'*4, 'Console Start', '>' * 4);
sample = int(input('Enter number of Sample(0 to end console): '));
length = int(input('Enter DNA Sequence length: '));

while(sample > 0):
    motif = {};
    takeDNASequence(sample , length);
    lmar = int(input('\nEnter l-mar(size of motif): '));
    Motif_Finding(lmar , sample);
    sample = int(input('\nEnter number of Sample(0 to end console): '));
    if(sample == 0):
        break;
    length = int(input('\nEnter DNA Sequence length: '));
print(">"*4 , 'End console' , '<'*4);

File: Bio_Informatics_CPP/Task-12/Motif Finding.cpp
#include<bits/stdc++.h>
using namespace std;

const int high = 100;

char DNA[4] = {'A' , 'T' , 'C' , 'G'};

vector<char>motif[high];
vector<string>getmotifs;

int check[high];

void clr(int length)
{
    for(int i=0; i<length; i++)
    {
        motif[i].clear();
        check[i] = 0;
    }
}

void takeDNAsequence(int row , int col)
{
    int i=0 , j=0;

    for(i=0; i<row; i++)
    {
        for(j=0; j<col; j++)
        {
            int indx = (rand() % 4);
            char ch = DNA[indx];
            motif[i].push_back(ch);
        }
    }

    cout << "\nDNA Sequences:\n";

    for(i=0; i<row; i++)
    {
        //cout << "for " << i << ":\n";

        for(j=0; j<motif[i].size(); j++)
        {
            cout << motif[i][j] << " ";
        }

        cout << "\n";
    }
}

void Motif_Finding(int row, int col , int lmar)
{
    int i=0 , j=0 , k=0, l=0, p=0;
    string s = "";
    bool fl=true;

    for(i=0; i<motif[0].size()-lmar+1; i++)
    {
        s = "";

        for(j=i; j<lmar+i; j++)
        {
            s += motif[0][j];
        }

        //cout << " s = " << s << "\n";

        string instring = "";

        for(k=1; k<row; k++)
        {
            instring = "";

            for(l=0; l<motif[k].size()-lmar+1; l++)
            {
                instring = "";

                for(p=l; p<lmar+l; p++)
                {
                    instring+=motif[k][p];
                }

                //cout << " instring = " << instring << "\n";

                if(s == instring)
                {
                    //cout << "find motif = " << s << "\n";

                    getmotifs.push_back(instring);

                    check[k] = 1;
                    //cout << "k = " << k << "\n";
                }

                else
                {
                    continue;
                }
            }

            fl=true;

            for(int x=1; x<row; x++)
            {
                if(!check[x])
                {
                    fl=false;
                    break;
                }
            }

            if(fl)
            {
               break;
            }

            else
            {
                getmotifs.clear();

                for(int h=0; h<high; h++) check[h] = 0;
            }
        }

        if(fl) break;
    }

    if(fl)
    {
        cout << "\n>> Motif:\n";

        for(i=0; i<getmotifs.size(); i++) cout << getmotifs[i] << "\n";
    }

    else
    {
        cout << "\n<< Motif did not found\n";
    }
}

int main()
{
    int length, lmar , sample;
    cout << "Enter Sample(0 to end terminal): ";
    cin >> sample;
    cout << "\nEnter DNA Sequence Length: ";
    cin >> length;
    while(sample)
    {
        clr(max(sample , length));
        getmotifs.clear();

        takeDNAsequence(sample , length);

        cout << "\nEnter l-mar: ";
        cin >> lmar;

        Motif_Finding(sample , length , lmar);

        cout << "\n\nEnter Sample(0 to end terminal): ";
        cin >> sample;
        if(!sample) break;
        cout << "\nEnter DNA Sequence Length(0 to end terminal): ";
        cin >> length;
        if(!length) break;
    }

    cout << "\n<<<< End terminal >>>>>\n";

    return 0;
}
